From a092ebbd0e4501de45778f20af6b298401aec944 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 30 Nov 2023 17:29:36 +0100
Subject: [PATCH 0184/1306] winex11: Attach client windows after their
 creation.

We still create them with whole_window parent when it exists as
otherwise Steam Overlay will completely mess up later on.

CW-Bug-Id: #23195
---
 dlls/winex11.drv/window.c | 33 +++++++++++++++------------------
 1 file changed, 15 insertions(+), 18 deletions(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 8430c08f712..f7b560fd0ca 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1771,10 +1771,10 @@ Window create_dummy_client_window(void)
  */
 Window create_client_window( HWND hwnd, const XVisualInfo *visual, Colormap colormap )
 {
-    Window dummy_parent = get_dummy_parent();
     struct x11drv_win_data *data = get_win_data( hwnd );
+    Window parent_window, client_window;
     XSetWindowAttributes attr;
-    Window ret;
+    unsigned int attr_mask;
     int x, y, cx, cy;
 
     if (!data)
@@ -1787,38 +1787,35 @@ Window create_client_window( HWND hwnd, const XVisualInfo *visual, Colormap colo
         data->window_rect = data->whole_rect = data->client_rect;
     }
 
-    detach_client_window( data, data->client_window, TRUE );
-
     attr.colormap = colormap;
     attr.bit_gravity = NorthWestGravity;
     attr.win_gravity = NorthWestGravity;
     attr.backing_store = NotUseful;
     attr.border_pixel = 0;
+    attr_mask = CWColormap | CWBitGravity | CWWinGravity | CWBackingStore | CWBorderPixel;
 
     x = data->client_rect.left - data->whole_rect.left;
     y = data->client_rect.top - data->whole_rect.top;
     cx = min( max( 1, data->client_rect.right - data->client_rect.left ), 65535 );
     cy = min( max( 1, data->client_rect.bottom - data->client_rect.top ), 65535 );
 
-    ret = data->client_window = XCreateWindow( gdi_display,
-                                               data->whole_window ? data->whole_window : dummy_parent,
-                                               x, y, cx, cy, 0, default_visual.depth, InputOutput,
-                                               visual->visual, CWBitGravity | CWWinGravity |
-                                               CWBackingStore | CWColormap | CWBorderPixel, &attr );
-    if (data->client_window)
+    /* NOTE: Creating the client windows as child of dummy parent has bad interactions with
+     * Steam Overlay and will cause whole_window to minimize when the overlay opens... */
+    parent_window = data->whole_window ? data->whole_window : get_dummy_parent();
+    if ((client_window = XCreateWindow( gdi_display, parent_window, x, y, cx, cy, 0, default_visual.depth,
+                                        InputOutput, visual->visual, attr_mask, &attr )))
     {
-        XSaveContext( data->display, data->client_window, winContext, (char *)data->hwnd );
+        XFlush( gdi_display ); /* make sure client_window is created for XSelectInput */
+        XSync( data->display, False ); /* make sure client_window is known from data->display */
+
+        attach_client_window( data, client_window );
         XMapWindow( gdi_display, data->client_window );
-        if (data->whole_window)
-        {
-            XFlush( gdi_display ); /* make sure client_window is created for XSelectInput */
-            XSync( data->display, False ); /* make sure client_window is known from data->display */
-            XSelectInput( data->display, data->client_window, ExposureMask );
-        }
+
         TRACE( "%p xwin %lx/%lx\n", data->hwnd, data->whole_window, data->client_window );
     }
+
     release_win_data( data );
-    return ret;
+    return client_window;
 }
 
 
-- 
2.43.2

